gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java

```diff
@@ -685,8 +674,23 @@ public void apply(JobListener jobListener, JobContext jobContext)
         }
       }
     } finally {
-      // Stop metrics reporting
+      // Register metrics then stop metrics reporting, metrics will flush and emit when the metricsContext is closed
       if (this.jobContext.getJobMetricsOptional().isPresent()) {
+        if (jobState.getPropAsBoolean(ConfigurationKeys.GOBBLIN_OUTPUT_JOB_LEVEL_METRICS, true)) {
+          String workunitCreationGaugeName = MetricRegistry.name(ServiceMetricNames.GOBBLIN_JOB_METRICS_PREFIX,
+              TimingEvent.LauncherTimings.WORK_UNITS_CREATION, jobState.getJobName());
+          long workUnitsCreationTime = workUnitsCreationTimer.getDuration() / TimeUnit.SECONDS.toMillis(1);
```

Comment:
all that's here seems potentially excessive for a `finally` block.  in general, it's better reserved for managing resource lifecycle, rather than to perform a bunch of initialization/calculation that may itself throw an exception (when `finally` could already be unwinding from one).

some specific Qs:
a. is there a reason, say timing imprecision when calling `getDuration()`, not to keep within the `try`?
b. does the event really need to be submitted even if there was an error/exception?

Reply:
Yeah there might be a better method of handling the metrics rather in finally, but I do think it is important to always emit metrics for each execution of a pipeline hence why trying to consolidate them all within a finally block. Maybe the solution here is to add yet another `try` in the `finally`? Reason being is that users should also have visibility when their workflow does not perform work, rather than not receiving information at all which would not activate alerting.

Reply:
I agree that we should always emit a workunit count because this is most useful when the workflow does _not_ perform work. Wrapping it in a try can make this safer to avoid throwing an exception.